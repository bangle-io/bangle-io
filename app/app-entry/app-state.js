import { naukarWorkerProxy } from 'naukar-proxy';
import { objectSync } from 'object-sync/index';
import * as Comlink from 'comlink';
import { validateNonWorkerGlobalScope } from 'naukar-worker';
import { initialAppState } from 'shared/index';
import { moduleSupport } from 'worker-setup/index';

validateNonWorkerGlobalScope();

export const appState = objectSync(initialAppState, (event) => {
  naukarWorkerProxy.updateAppState(event);
});

if (moduleSupport) {
  naukarWorkerProxy.registerUpdateCallback(
    Comlink.proxy((event) => appState.applyChange(event)),
  );
} else {
  naukarWorkerProxy.registerUpdateCallback((event) =>
    appState.applyChange(event),
  );
}
