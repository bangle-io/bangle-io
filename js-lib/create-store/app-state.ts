import type { JsonObject, JsonValue } from 'type-fest';

import type { BaseAction, Slice, SliceStateField } from './app-state-slice';

class AppStateConfig<S, A extends BaseAction, Op> {
  slices: SliceArray<S, A> = [];
  slicesByKey: { [k: string]: Slice<any, A, S> } = Object.create(null);
  fields: Array<FieldDesc<S, A, Op>> = [];
  opts?: Op;

  constructor(slices: SliceArray<S, A>, opts?: Op) {
    this.opts = opts;

    slices.forEach((slice) => {
      if (this.slicesByKey[slice.key]) {
        throw new RangeError(
          'Adding different instances of an existing slice (' + slice.key + ')',
        );
      }

      this.slices.push(slice);
      this.slicesByKey[slice.key] = slice;

      if (slice.spec.state) {
        this.fields.push(new FieldDesc(slice.key, slice.spec.state, slice));
      }
    });
  }
}

export class AppState<S = any, A extends BaseAction = any, Op = any> {
  static create<S, A extends BaseAction, Op = any>({
    slices,
    opts,
  }: {
    slices: SliceArray<S, A>;
    opts?: Op;
  }): AppState<S, A, Op> {
    const config = new AppStateConfig(slices, opts);
    const instance = new AppState(config);

    config.fields.forEach((f) => {
      instance.slicesCurrentState[f.name] = f.init(config.opts, instance);
    });

    return instance;
  }

  static stateFromJSON<S, A extends BaseAction = any, Op = any>({
    slices,
    json,
    // an object with a unique name for each
    // slice used from uniquely identifying a
    // slices data.
    sliceFields,
    opts,
  }: {
    slices: SliceArray<S, A>;
    json: JsonObject;
    sliceFields: { [key: string]: Slice };
    opts?: Op;
  }) {
    const config = new AppStateConfig(slices, opts);
    const instance = new AppState(config);

    config.fields.forEach((f) => {
      for (var prop in sliceFields) {
        const slice = sliceFields[prop];
        const state = slice?.spec.state;

        if (
          slice &&
          slice.key === f.name &&
          state &&
          state.stateFromJSON &&
          Object.prototype.hasOwnProperty.call(json, prop)
        ) {
          let val = json[prop];
          // This field belongs to a plugin mapped to a JSON field, read it from there.
          instance.slicesCurrentState[f.name] = state.stateFromJSON.call(
            slice,
            config,
            val!,
            instance,
          );

          return;
        }
      }
      instance.slicesCurrentState[f.name] = f.init(config.opts, instance);
    });

    return instance;
  }

  protected slicesCurrentState: { [k: string]: any } = Object.create(null);
  constructor(public config: AppStateConfig<S, A, Op>) {}
  applyAction(rootAction: A): AppState<S, A, Op> {
    let newInstance = this._applyInner(rootAction);

    let actions: A[] = [rootAction];
    let seen: Array<{ state: AppState; n: number }> | null = null;

    for (;;) {
      let haveNew = false;
      let i = -1;
      for (const slice of this.config.slices) {
        i++;

        if (!slice.spec.appendAction) {
          continue;
        }

        let n = seen ? seen[i]!.n : 0;
        let newAction =
          n < actions.length &&
          slice.spec.appendAction.call(
            slice,
            n ? actions.slice(n) : actions,
            newInstance,
          );

        if (newAction) {
          newAction.appendedFrom = rootAction.name;

          if (!seen) {
            seen = [];
            for (let j = 0; j < this.config.slices.length; j++) {
              seen.push(
                j < i
                  ? { state: newInstance, n: actions.length }
                  : { state: this, n: 0 },
              );
            }
          }
          actions.push(newAction as A);
          newInstance = newInstance._applyInner(newAction);
          haveNew = true;
        }
        if (seen) {
          seen[i] = { state: newInstance, n: actions.length };
        }
      }

      if (!haveNew) {
        return newInstance;
      }
    }
  }

  getSliceByKey<SL, A extends BaseAction, S>(
    key: string,
  ): Slice<SL, A, S> | undefined {
    return this.config.slicesByKey[key] as any;
  }

  getSlices() {
    return this.config.slices;
  }

  getSliceState<SL>(key: string): SL | undefined {
    return this.slicesCurrentState[key];
  }

  stateToJSON({
    sliceFields,
  }: {
    sliceFields?: { [key: string]: Slice };
  }): JsonObject {
    let result: { [rec: string]: JsonValue } = {};
    for (var prop in sliceFields) {
      const slice = sliceFields[prop];
      const state = slice?.spec.state;

      if (state && state.stateToJSON) {
        result[prop] = state.stateToJSON.call(
          slice,
          this.getSliceState(slice.key)!,
        );
      }
    }

    return result;
  }

  private _applyInner<AA extends BaseAction>(action: AA): AppState<S, A, Op> {
    let newInstance = new AppState(this.config);

    this.config.fields.forEach((field) => {
      if (field.apply) {
        newInstance.slicesCurrentState[field.name] = field.apply(
          action,
          this.getSliceState(field.name),
          newInstance,
        );
      } else {
        newInstance.slicesCurrentState[field.name] = this.getSliceState(
          field.name,
        );
      }
    });

    return newInstance;
  }
}

// eslint-disable-next-line @typescript-eslint/ban-types
function bind(f?: Function, self?: object) {
  return !self || !f ? f : f.bind(self);
}

export type SliceArray<S, A extends BaseAction> = Array<Slice<any, A, S>>;

class FieldDesc<S, A extends BaseAction, Op> {
  init: (config: undefined | Op, appState: AppState<S, A, Op>) => any;

  apply?: (action: any, value: any, appState: AppState<S, A, Op>) => any;

  constructor(
    public name: string,
    desc: {
      init: SliceStateField<any, any, S>['init'];
      apply?: SliceStateField<any, any, S>['apply'];
    },
    self: SliceArray<S, A>[0],
  ) {
    this.init = bind(desc.init, self);

    if (desc.apply) {
      this.apply = bind(desc.apply, self);
    }
  }
}
