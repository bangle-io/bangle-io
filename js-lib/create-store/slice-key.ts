import { AppState } from './app-state';
import type {
  BaseAction,
  ExtractAction,
  Slice,
  SliceSideEffect,
} from './app-state-slice';
import { ApplicationStore } from './app-store';

const keys: { [k: string]: number } = Object.create(null);

export function createKey(name: string) {
  if (name in keys) {
    return name + '$' + ++keys[name];
  }
  keys[name] = 0;

  return name + '$';
}

export class SliceKey<
  SL = any,
  A extends BaseAction = any,
  S = SL,
  C extends { [key: string]: any } = any,
> {
  key: string;

  constructor(public name: string) {
    this.key = createKey(name);
  }

  // serialization type helpers
  actionSerializer<ANAME extends A['name'], T>(
    actionName: ANAME,
    // return the serialzed value of the action
    toJSON: (action: ExtractAction<A, ANAME>) => T,
    // return the parsed value of the action
    fromJSON: (serialActionValue: T) => ExtractAction<A, ANAME>['value'],
  ) {
    return {
      toJSON,
      fromJSON,
    };
  }

  // Helper function for creating an operation with the correct
  asyncOp<T extends Promise<any>>(
    cb: (
      state: AppState<S, A>,
      dispatch: ApplicationStore<SL, A>['dispatch'],
      store: ApplicationStore<SL, A>,
    ) => T,
  ) {
    return cb;
  }

  // return[1] - the first dependency name that change
  didChange<K extends keyof SL>(
    state: AppState<any, any> | Readonly<AppState<any, any>>,
    prevState: AppState<any, any> | Readonly<AppState<any, any>>,
  ) {
    return <T extends K[]>(...dependencies: [...T]) => {
      let changed = false;
      let fieldChanged: K | undefined = undefined;

      dependencies.forEach((field) => {
        if (changed === false) {
          changed = this.valueChanged(field, state, prevState);
          if (changed === true) {
            fieldChanged = field;
          }
        }
      });

      return [changed, fieldChanged];
    };
  }

  // Helper function create a side effect with the correct type.
  effect(cb: SliceSideEffect<SL, A, C>) {
    return cb;
  }

  getDispatch(
    dispatch: ApplicationStore<any, any>['dispatch'],
  ): ApplicationStore<SL, A>['dispatch'] {
    return dispatch as ApplicationStore<SL, A>['dispatch'];
  }

  getSlice(
    state: AppState<S, A> | Readonly<AppState<S, A>>,
  ): Slice<SL, A, S, C> | undefined {
    return state.getSliceByKey(this.key);
  }

  getSliceState(
    state: AppState<any, any> | Readonly<AppState<any, any>>,
  ): SL | undefined {
    return state.getSliceState(this.key);
  }

  getSliceStateAsserted(
    state: AppState<any, any> | Readonly<AppState<any, any>>,
  ): SL {
    const sliceState: SL | undefined = state.getSliceState(this.key);
    if (sliceState === undefined) {
      throw new Error(`Slice state for "${this.key}"" cannot be undefined`);
    }

    return sliceState;
  }

  getState(
    storeOrState: ApplicationStore | AppState,
  ): ApplicationStore<SL, A>['state'] {
    if (storeOrState instanceof AppState) {
      return storeOrState;
    }

    return this.getStore(storeOrState).state;
  }

  // is a type helper to make it easy calling external operations
  getStateAndDispatch(store: ApplicationStore) {
    return {
      state: this.getState(store),
      dispatch: this.getDispatch(store.dispatch),
    };
  }

  // while keeping TS happy.
  getStore(store: ApplicationStore): ApplicationStore<SL, A> {
    return store;
  }

  //  use `valueChanged` if you expect `undefined` to be a valid value.
  getValueIfChanged<T extends keyof SL>(
    field: T,
    state: AppState<any, any> | Readonly<AppState<any, any>>,
    prevState: AppState<any, any> | Readonly<AppState<any, any>>,
  ): SL[T] | undefined {
    return this.valueChanged(field, state, prevState)
      ? this.getSliceStateAsserted(state)[field]
      : undefined;
  }

  // types.
  op<T>(
    cb: (
      state: AppState<S, A>,
      dispatch: ApplicationStore<SL, A>['dispatch'],
    ) => T,
  ) {
    return cb;
  }

  // types.
  queryOp<T>(cb: (state: AppState<S, A>) => T) {
    return cb;
  }

  valueChanged(
    field: keyof SL,
    state: AppState<any, any> | Readonly<AppState<any, any>>,
    prevState: AppState<any, any> | Readonly<AppState<any, any>>,
  ): boolean {
    return (
      this.getSliceStateAsserted(state)[field] !==
      this.getSliceStateAsserted(prevState)[field]
    );
  }

  // gets the value if it was different from prevState
  // WARNING! it will return undefined if the value is the same
  //  which can be problematic if your field can have `undefined`.

  // Similar to getValueIfChanged but instead takes a list of dependencies
  // and returns a two tuple of whether there was any change and the first field that changed.
  // return[0] - true if any of them changed

  // Helper function for creating an operation with the correct
}
