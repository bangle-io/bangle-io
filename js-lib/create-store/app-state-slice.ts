import type { JsonValue } from 'type-fest';

import type { AppState } from './app-state';
import type { ApplicationStore } from './app-store';
import type { SliceKey } from './slice-key';
import { createKey } from './slice-key';

export type BaseAction =
  | undefined
  | {
      name: string;
    };

// A - action
// SL - Slice's state
// S - AppState
export interface SliceStateField<SL, A extends BaseAction, S> {
  init(
    this: Slice<SL, A, S>,
    config: { [key: string]: any },
    appState: AppState<S, A>,
  ): SL;
  apply?(
    this: Slice<SL, A, S>,
    action: A,
    value: SL,
    appState: AppState<S, A>,
  ): SL;

  stateToJSON?: (this: Slice<SL, A, S>, value: SL) => JsonValue;
  stateFromJSON?: (
    this: Slice<SL, A, S>,
    config: { [key: string]: any },
    value: JsonValue,
    appState: AppState<S, A>,
  ) => SL;
}

export type SliceSideEffect<SL, A extends BaseAction, S = SL> = (
  store: ApplicationStore<S, A>,
) => {
  update?: (
    store: ApplicationStore<S, A>,
    prevState: AppState<S, A>,
    sliceState: SL,
    prevSliceState: SL,
  ) => void;
  destroy?: () => void;
  deferredUpdate?: (
    store: ApplicationStore<S, A>,
    abortSignal: AbortSignal,
  ) => void;
};

export class Slice<SL, A extends BaseAction = any, S = SL> {
  key: string;

  constructor(
    public spec: {
      key?: SliceKey<SL, A, S>;
      state?: SliceStateField<SL, A, S>;
      // false if it cannot be serialized
      actionToJSON?: (action: A & { id: string }) => string | false;
      actionFromJSON?: (jsonAction: string) => A;
      sideEffect?: SliceSideEffect<SL, A, S> | SliceSideEffect<SL, A, S>[];
    },
  ) {
    this.key = spec.key ? spec.key.key : createKey('slice');
  }

  getSliceState(state: AppState<S, A>): SL | undefined {
    return state.getSliceState(this.key);
  }
}
