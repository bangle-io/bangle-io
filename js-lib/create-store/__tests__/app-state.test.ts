import { AppState } from '../app-state';
import { Slice } from '../app-state-slice';
import { SliceKey } from '../slice-key';

describe('AppState', () => {
  test('empty slices', () => {
    const appState = AppState.create({ slices: [] });

    expect(appState).toMatchInlineSnapshot(`
      AppState {
        "config": AppStateConfig {
          "fields": [],
          "opts": undefined,
          "slices": [],
          "slicesByKey": {},
        },
        "slicesCurrentState": {},
      }
    `);
  });

  test('with a slice', () => {
    const slice = new Slice({
      state: {
        init: () => null,
      },
    });

    const appState = AppState.create({ slices: [slice] });

    expect(appState.getSliceState('slice$')).toBe(null);
    expect(appState.getSliceByKey('slice$')).toBe(slice);
    expect(appState).toMatchSnapshot();
  });

  test('using same key throws error', () => {
    const key1 = new SliceKey<null>('one');

    expect(() =>
      AppState.create({
        slices: [
          new Slice({
            key: key1,
            state: {
              init: () => null,
            },
          }),
          new Slice({
            key: key1,
            state: {
              init: () => null,
            },
          }),
        ],
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Adding different instances of an existing slice (one$)"`,
    );
  });

  test('applying action preserves states of those who donot have apply', () => {
    const key1 = new SliceKey<number>('one');

    const slice = new Slice({
      key: key1,
      state: {
        init: () => 0,
      },
    });

    const appState = AppState.create({ slices: [slice] });
    expect(key1.getSliceState(appState)).toBe(0);

    let newAppState = appState.applyAction({ name: 'my-action' });
    expect(key1.getSliceState(newAppState)).toBe(0);
  });

  test('applying action preserves states for  apply which returns prev state', () => {
    let initialValue = {};
    const key1 = new SliceKey<{}>('one');

    const slice = new Slice({
      key: key1,
      state: {
        init: () => initialValue,
        apply: (_, val) => val,
      },
    });

    const appState = AppState.create({ slices: [slice] });
    expect(key1.getSliceState(appState)).toBe(initialValue);

    let newAppState = appState.applyAction({ name: 'my-action' });
    expect(key1.getSliceState(newAppState)).toBe(initialValue);
  });

  test('calls state methods correctly', () => {
    const key1 = new SliceKey<number>('one');

    let init = jest.fn(() => {
      return 0;
    });
    let apply = jest.fn(() => 1);
    const slice = new Slice({
      key: key1,
      state: {
        init,
        apply,
      },
    });

    const appState = AppState.create({ slices: [slice], opts: { kj: true } });
    expect(init).toBeCalledTimes(1);
    expect(apply).toBeCalledTimes(0);
    expect(init).nthCalledWith(1, { kj: true }, appState);

    let action = {
      name: 'for-b',
      value: { no: 77 },
    };
    let newAppState = appState.applyAction(action);
    expect(init).toBeCalledTimes(1);
    expect(apply).toBeCalledTimes(1);
    expect(apply).nthCalledWith(1, action, 0, newAppState);
  });

  test('calls with a partial state', () => {
    const sliceKeyA = new SliceKey('a');
    const sliceKeyB = new SliceKey('b');

    let partialStateCheck;

    const sliceA = new Slice({
      key: sliceKeyA,
      state: {
        init: () => 1,
        apply: (action, value, appState) => {
          if (action.name === 'for-a') {
            return action.value.no;
          }

          return value;
        },
      },
    });

    const sliceB = new Slice({
      key: sliceKeyB,
      state: {
        init: () => 2,
        apply: (action, value, appState) => {
          partialStateCheck = sliceA.getSliceState(appState);

          return value;
        },
      },
    });

    const appState = AppState.create({ slices: [sliceA, sliceB] });

    let action = {
      name: 'for-a',
      value: { no: 77 },
    };
    appState.applyAction(action);
    // Should have access to A's state
    expect(partialStateCheck).toBe(77);
  });

  test('called with the right this', () => {
    const key1 = new SliceKey<number>('one');
    let initThis;
    let applyThis;
    const slice = new Slice({
      key: key1,
      state: {
        init() {
          initThis = this;

          return 0;
        },
        apply(action, val) {
          applyThis = this;

          return val;
        },
      },
    });

    const appState = AppState.create({ slices: [slice] });

    let action = {
      name: 'for-b',
      value: { no: 77 },
    };
    appState.applyAction(action);

    expect(initThis).toBe(slice);
    expect(applyThis).toBe(slice);
  });

  test('with multiple slices', () => {
    const key1 = new SliceKey<number>('one');
    const slice1 = new Slice({
      key: key1,
      state: {
        init: () => 1,
      },
    });
    const key2 = new SliceKey<number>('two');

    const slice2 = new Slice({
      key: key2,
      state: {
        init: () => 2,
      },
    });

    const appState = AppState.create({ slices: [slice1, slice2] });

    expect(key1.getSliceState(appState)).toBe(1);
    expect(key1.getSlice(appState)).toBe(slice1);
    expect(appState.getSliceState(key1.key)).toBe(1);
    expect(appState.getSliceByKey(key1.key)).toBe(slice1);

    expect(key2.getSliceState(appState)).toBe(2);
    expect(key2.getSlice(appState)).toBe(slice2);
    expect(appState.getSliceState(key2.key)).toBe(2);
    expect(appState.getSliceByKey(key2.key)).toBe(slice2);

    expect(appState).toMatchSnapshot();
  });

  test('updating a slice', () => {
    const sliceKey = new SliceKey('one');
    const slice = new Slice({
      key: sliceKey,
      state: {
        init: () => 1,
        apply: (_, val) => val + 1,
      },
    });

    let appState = AppState.create({ slices: [slice] });

    expect(sliceKey.getSliceState(appState)).toBe(1);

    let newAppState = appState.applyAction({ name: 'hi' });

    expect(newAppState).not.toBe(appState);

    expect(sliceKey.getSliceState(appState)).toBe(1);
    expect(sliceKey.getSliceState(newAppState)).toBe(2);
  });

  test('updating multiple slice', () => {
    const sliceKeyA = new SliceKey<number, ActionType>('a');
    const sliceKeyB = new SliceKey<number, ActionType>('b');

    type ActionType =
      | {
          name: 'for-a';
          value: { n: number };
        }
      | {
          name: 'for-b';
          value: { n: number };
        };

    const sliceA = new Slice({
      key: sliceKeyA,
      state: {
        init: () => 1,
        apply: (action, value, appState) => {
          if (action.name === 'for-a') {
            return action.value.n;
          }

          return value;
        },
      },
    });

    const sliceB = new Slice({
      key: sliceKeyB,
      state: {
        init: () => 2,
        apply: (action, value, appState) => {
          if (action.name === 'for-b') {
            return action.value.n;
          }

          return value;
        },
      },
    });

    let appState = AppState.create({ slices: [sliceA, sliceB] });

    expect(sliceKeyA.getSliceState(appState)).toBe(1);

    let newAppState = appState.applyAction({
      name: 'for-a',
      value: { n: 99 },
    });

    expect(newAppState).not.toBe(appState);
    // check new state
    expect(sliceKeyA.getSliceState(newAppState)).toBe(99);
    expect(sliceKeyB.getSliceState(newAppState)).toBe(2);

    // check old state
    expect(sliceKeyA.getSliceState(appState)).toBe(1);
    expect(sliceKeyB.getSliceState(appState)).toBe(2);

    newAppState = newAppState.applyAction({
      name: 'for-b',
      value: { n: 77 },
    });

    expect(sliceKeyA.getSliceState(newAppState)).toBe(99);
    expect(sliceKeyB.getSliceState(newAppState)).toBe(77);
  });

  describe('serialization', () => {
    test('serializes correctly', () => {
      const key1 = new SliceKey<any>('one');

      const slice1 = new Slice({
        key: key1,
        state: {
          init: () => ({ number: 3 }),
          stateToJSON: (val) => val.number,
          stateFromJSON: (config, val, appState) => ({ number: val }),
        },
      });

      const slice2 = new Slice({
        state: {
          init: () => 4,
        },
      });

      let state = AppState.create({
        slices: [slice1, slice2],
      });

      const sliceFields = { myslice1: slice1, myslice2: slice2 };
      const json = state.stateToJSON({ sliceFields: sliceFields });
      expect(json).toMatchInlineSnapshot(`
        {
          "myslice1": 3,
        }
      `);

      const parsedState = AppState.stateFromJSON({
        slices: [slice1, slice2],
        json,
        sliceFields,
      });
      expect(slice1.getSliceState(parsedState)).toEqual({ number: 3 });
      expect(slice2.getSliceState(parsedState)).toBe(4);
    });

    test('skips serialization if no field name provided', () => {
      const key1 = new SliceKey<any>('one');

      const slice1 = new Slice({
        key: key1,
        state: {
          init: () => ({ number: 3 }),
          stateToJSON: (val) => val.number,
          stateFromJSON: (config, val, appState) => ({ number: val }),
        },
      });

      let state = AppState.create({
        slices: [slice1],
      });

      const sliceFields = {};
      const json = state.stateToJSON({ sliceFields: sliceFields });
      expect(json).toMatchInlineSnapshot(`{}`);

      const parsedState = AppState.stateFromJSON({
        slices: [slice1],
        json,
        sliceFields,
      });
      expect(slice1.getSliceState(parsedState)).toEqual({ number: 3 });
    });
  });

  describe('append action', () => {
    test('calls apply correctly', () => {
      const key1 = new SliceKey<number>('one');

      let apply = jest.fn((_, val) => val + 1);

      const slice = new Slice({
        key: key1,
        state: {
          init: () => 0,
          apply: apply,
        },
        appendAction(actions) {
          if (actions.some((a) => a.name === 'for-b')) {
            return {
              name: 'for-b-appended',
            };
          }

          return undefined;
        },
      });

      const appState = AppState.create({ slices: [slice] });
      expect(apply).toBeCalledTimes(0);

      let action = {
        name: 'for-b',
        value: { no: 77 },
      };

      let newAppState = appState.applyAction(action);

      expect(apply).toBeCalledTimes(2);

      expect(apply).nthCalledWith(1, action, 0, expect.any(AppState));
      expect(apply.mock.calls[1]?.[0]).toEqual({
        appendedFrom: 'for-b',
        name: 'for-b-appended',
      });
      expect(apply.mock.calls[1]?.[1]).toEqual(1);

      // should not have any effect
      appState.applyAction({
        name: 'something-else',
        value: { no: 77 },
      });

      expect(key1.getSliceState(newAppState)).toEqual(2);
      expect(apply).toBeCalledTimes(3);
    });

    test('fund test', () => {
      const key1 = new SliceKey<string>('one');

      const slice1 = new Slice({
        key: key1,
        state: {
          init: () => '👻',
          apply: (action, value) => value + action.value.emoji,
        },
        appendAction(actions, state) {
          if (actions.some((a) => a.name.startsWith('interesting-action'))) {
            return {
              name: 'appended-action',
              value: { emoji: '💩' },
            };
          }

          return undefined;
        },
      });

      const state = AppState.create({ slices: [slice1] });
      const newState = state.applyAction({
        name: 'interesting-action',
        value: {
          emoji: '🐔',
        },
      });

      expect(key1.getSliceState(newState)).toMatchInlineSnapshot(`"👻🐔💩"`);
    });

    test('multiple appends', () => {
      const key1 = new SliceKey<string>('one');
      const key2 = new SliceKey<string>('two');
      const slice1Actions: any[] = [];

      const slice1 = new Slice({
        key: key1,
        state: {
          init: () => 'Slice1:',
          apply: (action, value) => value + action.value.data,
        },
        appendAction(actions): any {
          slice1Actions.push([...actions]);

          if (actions.some((a) => a.name.startsWith('2-'))) {
            return {
              name: '1-appended',
              value: { data: '1 ' },
            };
          }

          return undefined;
        },
      });

      let times = 0;

      const slice2Actions: any[] = [];
      const slice2 = new Slice({
        key: key2,
        state: {
          init: () => 'Slice2:',
          apply: (action, value) => value + action.value.data,
        },
        appendAction(actions): any {
          slice2Actions.push([...actions]);

          if (
            (actions.some((a) => a.name.startsWith('1-')) && times < 4) ||
            actions[0].name === 'start'
          ) {
            times++;

            return {
              name: '2-appended-' + times,
              value: { data: '2 ' },
            };
          }

          return undefined;
        },
      });

      const appState = AppState.create({ slices: [slice1, slice2] });
      let newState = appState.applyAction({
        name: 'start',
        value: { data: 'start ' },
      });

      expect(key1.getSliceState(newState)).toEqual(
        'Slice1:start 2 1 2 1 2 1 2 1 ',
      );
      expect(key2.getSliceState(newState)).toEqual(
        'Slice2:start 2 1 2 1 2 1 2 1 ',
      );

      expect(slice1Actions).toEqual([
        [
          {
            name: 'start',
            value: { data: 'start ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '2-appended-1',
            value: { data: '2 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '2-appended-2',
            value: { data: '2 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '2-appended-3',
            value: { data: '2 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '2-appended-4',
            value: { data: '2 ' },
          },
        ],
      ]);
      expect(slice2Actions).toEqual([
        [
          {
            name: 'start',
            value: { data: 'start ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '1-appended',
            value: { data: '1 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '1-appended',
            value: { data: '1 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '1-appended',
            value: { data: '1 ' },
          },
        ],
        [
          {
            appendedFrom: 'start',
            name: '1-appended',
            value: { data: '1 ' },
          },
        ],
      ]);
    });
  });
});
