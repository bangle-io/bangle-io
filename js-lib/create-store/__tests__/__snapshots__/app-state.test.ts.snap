// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppState with a slice 1`] = `
AppState {
  "config": AppStateConfig {
    "fields": [
      FieldDesc {
        "apply": undefined,
        "init": [Function],
        "name": "slice$",
      },
    ],
    "opts": undefined,
    "slices": [
      Slice {
        "key": "slice$",
        "spec": {
          "state": {
            "init": [Function],
          },
        },
      },
    ],
    "slicesByKey": {
      "slice$": Slice {
        "key": "slice$",
        "spec": {
          "state": {
            "init": [Function],
          },
        },
      },
    },
  },
  "slicesCurrentState": {
    "slice$": null,
  },
}
`;

exports[`AppState with multiple slices 1`] = `
AppState {
  "config": AppStateConfig {
    "fields": [
      FieldDesc {
        "apply": undefined,
        "init": [Function],
        "name": "one$5",
      },
      FieldDesc {
        "apply": undefined,
        "init": [Function],
        "name": "two$",
      },
    ],
    "opts": undefined,
    "slices": [
      Slice {
        "key": "one$5",
        "spec": {
          "key": SliceKey {
            "key": "one$5",
            "name": "one",
          },
          "state": {
            "init": [Function],
          },
        },
      },
      Slice {
        "key": "two$",
        "spec": {
          "key": SliceKey {
            "key": "two$",
            "name": "two",
          },
          "state": {
            "init": [Function],
          },
        },
      },
    ],
    "slicesByKey": {
      "one$5": Slice {
        "key": "one$5",
        "spec": {
          "key": SliceKey {
            "key": "one$5",
            "name": "one",
          },
          "state": {
            "init": [Function],
          },
        },
      },
      "two$": Slice {
        "key": "two$",
        "spec": {
          "key": SliceKey {
            "key": "two$",
            "name": "two",
          },
          "state": {
            "init": [Function],
          },
        },
      },
    },
  },
  "slicesCurrentState": {
    "one$5": 1,
    "two$": 2,
  },
}
`;
